ko/resolve/resolve.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2018 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolve
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/google/go-containerregistry/ko/build"
+	"github.com/google/go-containerregistry/ko/publish"
+)
+
+// ImageReferences resolves supported references to images within the input yaml
+// to published image digests.
+func ImageReferences(input []byte, builder build.Interface, publisher publish.Interface) ([]byte, error) {
+	// The loop is to support multi-document yaml files.
+	// This is handled by using a yaml.Decoder and reading objects until io.EOF, see:
+	// https://github.com/go-yaml/yaml/pull/301/files#diff-e51f89b39d976550831afb510a926be0R123
+	decoder := yaml.NewDecoder(bytes.NewBuffer(input))
+	buf := bytes.NewBuffer(nil)
+	encoder := yaml.NewEncoder(buf)
+	for {
+		var obj interface{}
+		if err := decoder.Decode(&obj); err != nil {
+			if err == io.EOF {
+				return buf.Bytes(), nil
+			}
+			return nil, err
+		}
+
+		// Reflectively walk input, building and publishing supported references.
+		// TODO(mattmoor): It may be worth considering gathering the supported references in
+		// a first pass, performing the builds and pushes, and then performing the
+		// substitutions in a second pass.  This would enable us to eliminate the goroutine
+		// logic embedded in the recursive walk, naturally eliminate redundant build / publish
+		// for the same reference, and potentially enable us to create a single larger build
+		// (not clear if/how much this is possible in go build).
+		obj2, err := replaceRecursive(obj, func(ref string) (string, error) {
+			if !builder.IsSupportedReference(ref) {
+				return ref, nil
+			}
+			img, err := builder.Build(ref)
+			if err != nil {
+				return "", err
+			}
+			digest, err := publisher.Publish(img, ref)
+			if err != nil {
+				return "", err
+			}
+			return digest.String(), nil
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		if err := encoder.Encode(obj2); err != nil {
+			return nil, err
+		}
+	}
+}
+
+type replaceString func(string) (string, error)
+
+func replaceRecursive(obj interface{}, rs replaceString) (interface{}, error) {
```

Comment:
Can you add a comment describing what this does a little more. I'm struggling to parse this.

Reply:
Added a very detailed comment.